include/circt/Dialect/LLHD/IR/LLHDCanonicalization.td

```diff
@@ -0,0 +1,59 @@
+//===- LLHDCanonicalization.td - LLHD canonicalization -----*- tablegen -*-===//
+//
+// This file defines LLHD canonicalization patterns.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef CIRCT_DIALECT_LLHD_IR_CANONICALIZATION_TD
+#define CIRCT_DIALECT_LLHD_IR_CANONICALIZATION_TD
+
+include "circt/Dialect/LLHD/IR/LLHD.td"
+
+//===----------------------------------------------------------------------===//
+// XorOp
+//===----------------------------------------------------------------------===//
+
+def XorAllBitsSet : Pat<
+  (LLHD_XorOp $lhs, (LLHD_ConstOp AttrConstraint<
+    CPred<"$_self.isa<IntegerAttr>() && "
+      "$_self.cast<IntegerAttr>().getValue().isAllOnesValue()">,
+    "signless integer -1">)),
+  (LLHD_NotOp $lhs)>;
+
+//===----------------------------------------------------------------------===//
+// NotOp
+//===----------------------------------------------------------------------===//
+
+def NotOfEq : Pat<
+  (LLHD_NotOp (LLHD_EqOp $lhs, $rhs)),
+  (LLHD_NeqOp $lhs, $rhs)>;
+
+def NotOfNeq : Pat<
+  (LLHD_NotOp (LLHD_NeqOp $lhs, $rhs)),
+  (LLHD_EqOp $lhs, $rhs)>;
+
+//===----------------------------------------------------------------------===//
+// EqOp
+//===----------------------------------------------------------------------===//
+
+def BooleanEqToXor : Pat<
+  (LLHD_EqOp I1:$lhs, I1:$rhs),
+  (LLHD_NotOp (LLHD_XorOp $lhs, $rhs))>;
```

Comment:
I'm not sure I get what's the benefit of having a `not` of a `xor` instead of just the comparison for `i1` values?

Reply:
Boolean values are very commonly used in LLHD (e.g. for clk, enable, reset, etc.). The bitwise operations already have good folding and canonicalization. The goal is to take advantage of that. E.g. a common pattern I saw in LLHD is a bitwise operation with a `Eq` and a `Neq` as arguments. When compiling the accumulator example from the LLHD paper I got 
```mlir
^bb1: 
    %0 = llhd.prb %arg0 : !llhd.sig<i1>
    llhd.wait (%arg0 : !llhd.sig<i1>), ^bb2
^bb2: 
    %1 = llhd.prb %arg0 : !llhd.sig<i1>
    %false = constant false
    %2 = llhd.eq %0, %false : i1
    %3 = llhd.neq %1, %false : i1
    %4 = llhd.and %2, %3 : i1
```
which is transformed to the following by using this pattern
```mlir
^bb1:
    %1 = llhd.prb %arg0 : !llhd.sig<i1>
    llhd.wait (%arg0 : !llhd.sig<i1>), ^bb2
^bb2: 
    %2 = llhd.prb %arg0 : !llhd.sig<i1>
    %3 = llhd.not %1 : i1
    %4 = llhd.and %3, %2 : i1
```
Of course, I could also implement folders for `eq` and `neq` for this case, but then there is also the case where the equality operations cannot be folded, but the `and` can be folded because the equality operations have the same arguments. The options I see here would be to 
(1) add this pattern to every bitwise operation: makes bitwise folder more complex but also supports non-boolean values
(2) do nothing in this case
(3) convert to `xor` and take advantage of the complement folders of the bitwise operations: `and((eq a, b), (neq a, b) -> and((xor a, b), (not (xor a, b))) -> constant false`

Reply:
Thanks for the detailed explanation, sounds good!